File: src/Java/pipeline-core/src/main/java/com/pipeline/core/actions/MetricsOutputAction.java
package com.pipeline.core.actions;

import com.pipeline.core.ActionTiming;
import com.pipeline.core.StepAction;
import com.pipeline.core.StepControl;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.LongSupplier;

/**
 * Minimal out-of-the-box metrics as a post action (UBS-style):
 * measure per-action timings and emit a single metrics map to a sink.
 */
public final class MetricsOutputAction<C> implements StepAction<C> {
  private final String name;
  private final LongSupplier nanoClock;
  private final Consumer<Map<String, Object>> sink;

  public MetricsOutputAction() {
    this("Metrics", System::nanoTime, System.out::println);
  }

  public MetricsOutputAction(Consumer<Map<String, Object>> sink) {
    this("Metrics", System::nanoTime, sink);
  }

  public MetricsOutputAction(String name, Consumer<Map<String, Object>> sink) {
    this(name, System::nanoTime, sink);
  }

  public MetricsOutputAction(String name, LongSupplier nanoClock, Consumer<Map<String, Object>> sink) {
    this.name = Objects.requireNonNull(name, "name");
    this.nanoClock = Objects.requireNonNull(nanoClock, "nanoClock");
    this.sink = Objects.requireNonNull(sink, "sink");
  }

  @Override
  public C apply(C ctx, StepControl<C> control) {
    long nowNanos = nanoClock.getAsLong();
    long startNanos = control.runStartNanos();
    double pipelineLatencyMs = (startNanos == 0L) ? 0.0 : (nowNanos - startNanos) / 1_000_000.0;

    Map<String, Object> metricsMap = new LinkedHashMap<>();
    metricsMap.put("name", name);
    metricsMap.put("pipeline", control.pipelineName());
    metricsMap.put("shortCircuited", control.isShortCircuited());
    metricsMap.put("errorCount", control.errors().size());
    metricsMap.put("pipelineLatencyMs", pipelineLatencyMs);

    Map<String, Double> actionLatencyMs = new LinkedHashMap<>();
    for (ActionTiming timing : control.actionTimings()) {
      actionLatencyMs.put(timing.actionName(), timing.elapsedNanos() / 1_000_000.0);
    }
    metricsMap.put("actionLatencyMs", actionLatencyMs);

    sink.accept(metricsMap);
    return ctx;
  }
}

File: src/CSharp/pipeline_services/actions/MetricsOutputAction.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

using PipelineServices.Core;

namespace PipelineServices.Core.Actions;

public sealed class MetricsOutputAction<ContextType> : StepAction<ContextType>
{
    private readonly string name;
    private readonly Func<long> nanoClock;
    private readonly Action<IDictionary<string, object>> sink;

    public MetricsOutputAction()
        : this("Metrics", NanoTime.GetNowNanos, DefaultSink)
    {
    }

    public MetricsOutputAction(Action<IDictionary<string, object>> sink)
        : this("Metrics", NanoTime.GetNowNanos, sink)
    {
    }

    public MetricsOutputAction(string name, Action<IDictionary<string, object>> sink)
        : this(name, NanoTime.GetNowNanos, sink)
    {
    }

    public MetricsOutputAction(string name, Func<long> nanoClock, Action<IDictionary<string, object>> sink)
    {
        this.name = name ?? throw new ArgumentNullException(nameof(name));
        this.nanoClock = nanoClock ?? throw new ArgumentNullException(nameof(nanoClock));
        this.sink = sink ?? throw new ArgumentNullException(nameof(sink));
    }

    public ContextType Apply(ContextType contextValue, StepControl<ContextType> control)
    {
        if (control == null)
        {
            throw new ArgumentNullException(nameof(control));
        }

        long nowNanos = nanoClock();
        long startNanos = control.RunStartNanos();
        double pipelineLatencyMs = (startNanos == 0L) ? 0.0 : (nowNanos - startNanos) / 1_000_000.0;

        Dictionary<string, object> metricsMap = new Dictionary<string, object>(StringComparer.Ordinal);
        metricsMap["name"] = name;
        metricsMap["pipeline"] = control.PipelineName();
        metricsMap["shortCircuited"] = control.IsShortCircuited();
        metricsMap["errorCount"] = control.Errors().Count;
        metricsMap["pipelineLatencyMs"] = pipelineLatencyMs;

        Dictionary<string, double> actionLatencyMs = new Dictionary<string, double>(StringComparer.Ordinal);
        IReadOnlyList<ActionTiming> actionTimings = control.ActionTimings();
        for (int timingIndex = 0; timingIndex < actionTimings.Count; timingIndex++)
        {
            ActionTiming timing = actionTimings[timingIndex];
            actionLatencyMs[timing.ActionName] = timing.ElapsedNanos / 1_000_000.0;
        }
        metricsMap["actionLatencyMs"] = actionLatencyMs;

        sink(metricsMap);
        return contextValue;
    }

    private static void DefaultSink(IDictionary<string, object> metricsMap)
    {
        string json = JsonSerializer.Serialize(metricsMap);
        Console.WriteLine(json);
    }
}

